Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentEnd/FocusHelper.js

```diff
@@ -0,0 +1,25 @@
+/* vim: set ts=2 sts=2 sw=2 et tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+(function() {
+ "use strict";
+ const createFormElementEventHandler = (eventType) => {
```

Comment:
You shouldn't need to have this function to create the function for the event listener because the `event` that gets passed in already knows its type. I'm also not sure that we even need to filter *just* `<input>`/`<textarea>` or `[contenteditable]` elements since the `focus`/`blur` events shouldn't emit for anything that can't accept keyboard focus. You can just do:

```
function eventHandler(event) {
  const eventType = event.type;
  const elementType = event.target.nodeName;
  webkit.messageHandlers.focusHelper.postMessage({
    eventType, elementType
  });
}
```

Then, down below, you can just add these to `document` directly instead of `document.body`:

```
["focus", "blur"].forEach((eventType) => {
  document.addEventListener(eventType, eventHandler, options);
});
```

Reply:
Ah, good point re: `event.type`.

I had a look around what can emit focus and blur events. The standard seems a bit hazy, but [this from quirksmode][1]

 * `a` elements.
 * anything with a `tabindex` element.
 * the window itself.

As I'm using this as a proxy for "is the user editing text", I think I still need to test for the node name, and contenteditable.

[1]: https://www.quirksmode.org/dom/events/blurfocus.html

Reply:
Ah, right... `tabindex`... so yeah, just keep that bit to filter on `input,textarea,[contenteditable]`... actually, you could probably just do: `event.target.matches('input,textarea,[contenteditable]')

Reply:
Hmm.

If `contenteditable=false` or `contenteditable=inherit` does that still match?

I've added an `isButton` function which checks the buttonness too.